docs/dev/uap/implementation.rst

```diff
@@ -0,0 +1,333 @@
+  ``get_roles_for_collection`` and ``get_roles_for_user`` methods of
+  ``KolibriAbstractBaseUser``.
+
+Encoding Permission Rules
+-------------------------
+
+We need to associate a particular set of rules with each model, to specify the
+permissions that users should have in relation to instances of that model.
+While not all models have the same rules, there are some broad categories of
+models that do share the same rules (e.g. ContentInteractionLog,
+ContentSummaryLog, and UserSessionLog -- collectively, "User Log Data").
+Hence, it is useful to encapsulate a permissions "class" that can be reused
+across models, and extended (through inheritance) if slightly different
+behavior is needed. These classes of permissions are defined as Python classes
+that inherit from kolibri.auth.permissions.base.BasePermissions, which defines
+the following overridable methods:
+
+- The following four Boolean (True/False) permission checks, corresponding to
+  the "CRUD" operations:
+    - ``user_can_create_object``
+    - ``user_can_read_object``
+    - ``user_can_update_object``
+    - ``user_can_delete_object``
+- The queryset-filtering ``readable_by_user_filter`` method, which takes in a
+  queryset and returns a queryset filtered down to just objects that should be
+  readable by the user.
+
+Specifying Role-Based Permissions
+---------------------------------
+
+Defining a custom Permissions class and overriding its methods allows for
+arbitrary logic to be used in defining the rules governing the permissions,
+but many cases can be covered by more constrained rule specifications. In
+particular, the rules for many models can be specified in terms of the role-
+based permissions system described above. A built-in subclass of
+``BasePermissions``, called ``RoleBasedPermissions``, makes this easy.
+Creating an instance of ``RoleBasedPermissions`` involves defining the
+following parameters:
+
+- Lists of role kinds that should be granted each of the CRUD permissions,
+  encoded in the following parameters: ``can_be_created_by``, ``can_be_read_by``,
+  ``can_be_updated_by``, ``can_be_deleted_by``.
+- The ``target_field`` parameter that determines the "target" object for the
+  role-checking; this should be the name of a field on the model that foreign
+  keys either onto a ``FacilityUser`` or a ``Collection``. If the model we’re
+  checking permissions for is itself the target, then ``target_field`` may be
+  ``"."``.
+
+An example, showing that read permissions should be granted to a coach or
+admin for the user referred to by the model’s "user" field. Similarly, write
+permissions should only be available to an admin for the user::
+
+    class UserLogPermissions(RoleBasedPermissions):
+
+        can_be_created_by = ["admin"]
+        can_be_read_by = ["coach", "admin"]
+        can_be_updated_by = ["admin"]
+        can_be_deleted_by = ["admin"]
+
+        target_field = "user"
```

Comment:
This does not reflect how you have now implemented some permissions classes: https://github.com/learningequality/kolibri/pull/51/files#diff-efc2d6c2ce8481c7eee50b5837104665R10